proposals/gc/MVP.md

```diff
@@ -266,62 +293,112 @@ Tentatively, support a type of guaranteed unboxed scalars.
+  - `rtt.sub n $t1 $t2 : [(rtt n $t1)] -> [(rtt (n+1) $t2)]`
     - iff `t2 <: t1`
   - multiple invocations of this instruction with the same operand yield the same observable RTTs
   - this is a *constant instruction*
 
-TODO: Add the ability to generate new (non-canonical) RTT values to implement casting in nominal type hierarchies.
+TODO: Add the ability to generate new (non-canonical) RTT values to implement casting in nominal type hierarchies?
 
 
 #### Casts
 
-* `ref.test <heaptype1> <heaptype2>` tests whether a reference value's [runtime type](#values) is a [runtime subtype](#runtime) of a given RTT
-  - `ref.test t1 t2 : [(ref null t1) (rtt n t2)] -> [i32]`
-     - iff `t2 <: t1`
+RTT-based casts can only be performed with respect to concrete types, and require a data or function reference as input, which are known to carry an RTT.
+
+* `ref.test <typeidx>` tests whether a reference value's [runtime type](#values) is a [runtime subtype](#runtime) of a given RTT
+  - `ref.test $t : [(ref null ht) (rtt n $t)] -> [i32]`
+    - iff `ht <: data` or `ht <: func`
+    - and `(type $t) <: ht`
   - returns 1 if the first operand is not null and its runtime type is a sub-RTT of the RTT operand, 0 otherwise
 
-* `ref.cast <heaptype1> <heaptype2>` casts a reference value down to a type given by a RTT representation
-  - `ref.cast t1 t2 : [(ref null t1) (rtt n t2)] -> [(ref t2)]`
-     - iff `t2 <: t1`
+* `ref.cast <typeidx>` casts a reference value down to a type given by a RTT representation
+  - `ref.cast $t : [(ref null ht) (rtt n $t)] -> [(ref $t)]`
+    - iff `ht <: data` or `ht <: func`
+    - and `(type $t) <: ht`
   - traps if the first operand is null or its runtime type is not a sub-RTT of the RTT operand
 
-* `br_on_cast <labelidx> <heaptype1> <heaptype2>` branches if a value can be cast down to a given reference type
-  - `br_on_cast $l t1 t2 : [(ref null t1) (rtt n t2)] -> [(ref null t1)]`
-    - iff `t2 <: t1`
-    - and `$l : [(ref t2)]`
+* `br_on_cast <labelidx> <typeidx>` branches if a value can be cast down to a given reference type
+  - `br_on_cast $l $t : [(ref null ht) (rtt n $t)] -> [(ref null ht)]`
+    - iff `ht <: data` or `ht <: func`
+    - and `(type $t) <: ht`
+    - and `$l : [(ref $t)]`
   - branches iff the first operand is not null and its runtime type is a sub-RTT of the RTT operand
   - passes cast operand along with branch
 
+Note: The condition `(type $t) <: ht` isn't needed for soundness of any of the above instructions. If false, the check merely is statically known to fail. Should it be removed?
+
+
+#### Constant Expressions
+
+In order to allow RTTs to be initialised as globals, the following extensions are made to the definition of *constant expressions*:
+
+* `rtt.canon` is a constant instruction
+* `rtt.sub` is a constant instruction
+* `global.get` is a constant instruction and can access preceding global definitions, not just imports as in the MVP
```

Comment:
Preceding immutable global definitions.

Reply:
Yeah, changed for conservativeness's sake. Though I wonder if there is a hard reason for restricting it, given that no global can actually be mutated yet in that phase?